hw_5_pandas_code/hw_5_pandas.py

```diff
@@ -0,0 +1,239 @@
+# -*- coding: utf-8 -*-
+"""hw_5_pandas.ipynb
+
+Automatically generated by Colaboratory.
+
+Original file is located at
+    https://colab.research.google.com/drive/17dKL0EkWPPvBAgE0WeigvzYjzCk7m0oV
+"""
+
+import pandas as pd #requier
+import matplotlib.pyplot as plt
+import numpy as np
+import seaborn as sns #requier,uses pandas
+from matplotlib import rcParams
+
+"""## Задание 1. Работа с реальными данными"""
+
+def read_gff(path_to_file: str) -> pd.DataFrame:
+    return pd.read_csv(f'{path_to_file}', sep = '\t', comment='#',
+                       names = ['chromosome', 'source', 'type', 'start', 'end',
+                                'score', 'strand', 'phase', 'attributes'])
+
+rna_annotation = read_gff('data/rrna_annotation.gff')
+
+attribute_names = ['Name=16S_rRNA.*$', 'Name=23S_rRNA.*$','Name=5S_rRNA.*$']
+rna_annotation['attributes'] = rna_annotation['attributes'].replace(attribute_names, ['16S', '23S', '5S'],
+                                                                    regex=True)
+rna_annotation.head()
+
+def read_bed6(path_to_file: str) -> pd.DataFrame:
+    return pd.read_csv(f'{path_to_file}', sep='\t',
+                       names = ['chromosome', 'start', 'end',
+                                'name', 'score', 'strand'])
+
+alignment = read_bed6('data/alignment.bed')
+
+rna_annotation_tw_thr = rna_annotation[rna_annotation['attributes'] == '23S']
+s23_count = rna_annotation_tw_thr['chromosome'].value_counts()
+
+rna_annotation_six = rna_annotation[rna_annotation['attributes'] == '16S']
+s16_count = rna_annotation_six['chromosome'].value_counts()
+
+rna_annotation_five = rna_annotation[rna_annotation['attributes'] == '5S']
+s5_count = rna_annotation_five['chromosome'].value_counts()
+
+rna_count_table = pd.concat([s5_count.rename('5S'), s16_count.rename('16S'), s23_count.rename('23S')], axis = 1).sort_index()
+rna_count_table = rna_count_table.reset_index(level=0).rename({'index' : 'Sequence'}, axis = 'columns')
+
+#rna_final_table = pd.melt(rna_count_table, id_vars='Sequence', var_name='rRNA', value_name='Count')
+#rna_final_table.head()
+
+plt.style.use('default')
+rna_count_table.plot(x='Sequence', y=['16S', '23S', '5S'], kind='bar', figsize=(20, 10))
+plt.ylabel('Count')
+plt.legend(title='RNA type')
+
+rna_annotation['interval_ann'] = rna_annotation.apply(lambda row: (row.start, row.end), axis = 1)
+alignment['interval_align'] = alignment.apply(lambda row: (row.start - 1, row.end + 1), axis = 1)
+
+# Function to check if any two intervals overlap
+def intersect(first: tuple, second: tuple) -> bool:
+    if first[0] >= second[0] and first[1] <= second[1]:
+            return True
+    return False
+
+df_merge = rna_annotation.merge(alignment, on='chromosome')
+df_merge = df_merge[df_merge.apply(lambda row : intersect(row.interval_ann, row.interval_align), axis = 1)]
```

Comment:
Рабочий вариант, но можно ещё проще
```suggestion
df_merge.query("start_x >= start_y and end_x <= end_y")
```